lib/spack/spack/environment.py

```diff
@@ -845,6 +849,32 @@ def concretize(self, force=False):
             self.concretized_order = []
             self.specs_by_hash = {}
 
+        if self.concretize_together:
+            return self._concretize_together()
+
+        return self._concretize_separately()
+
+    def _concretize_together(self):
+        # Exit early if the set of concretized specs is the set of user specs
+        user_specs_did_not_change = not bool(
+            set(self.user_specs) - set(self.concretized_user_specs)
```

Comment:
> .user_specs are the specs that the user writes into the spack.yaml file, so they will almost always be different than .concretized_user_specs

@scheibelp Reading the code here:

https://github.com/spack/spack/blob/76b9c561109e4a19375231fd6e8fd8a1c7947f34/lib/spack/spack/environment.py#L977-L1000

it seems to me that `.concretized_user_specs` is the list of user specs that have been concretized already. The concrete specs goes into `.specs_by_hash` as a value.

So what I'm doing above is checking if in the `spack.yaml` file I have entries that were not concretized previously. If so I redo concretization of all the specs together, otherwise I return an empty list.

Reply:
Whoops you are definitely correct about that. This looks fine.